### ISSUE TITLE
[RFC] Loss functions in quickvision

### ISSUE THREAD

**Issue by oke-aditya** [Quick-AI/quickvision#43]
## 🚀 Feature
Frequently re-used losses that can be added to bolts.

### Motivation

Writing Losses is quite repetitive. PyTorch supports losses which are written with deep interoperability with C++ API. But most research losses aren't. Can these go to bolts/lightning (I feel it won't be good in lightning and better in bolts).

These losses are building blocks for other complicates losses as well.

### Pitch

A non-exhaustive and probable list of losses that are not PyTorch but used often.

- [ ] gIoU loss -> Used in Detr (will make porting Detr easier, present in [fvcore](https://github.com/facebookresearch/fvcore)
- [ ] focal loss -> Used in RetinaNet (It will come in torchvision as well but we we can have here too for re-use, present in [fvcore](https://github.com/facebookresearch/fvcore)
- [ ] Jsd loss -> Used alternative to CrossEntropy in Classification [here](https://github.com/rwightman/pytorch-image-models/blob/master/timm/loss/jsd.py)
- [ ] Varients of Cross Entropy -> Unsure if they are used often found them [here](https://github.com/rwightman/pytorch-image-models/blob/master/timm/loss/cross_entropy.py)
- [x] Dice Loss -> Used in U-Net and other segmentation models.
- [ ] Sigmoid Focal Loss -> Modification of Focal loss for segmentation.
- [ ] Huber loss -> Used in efficnet Det and similar loss. Implemented [here](https://github.com/rwightman/efficientdet-pytorch/blob/2fc4e0e107c969090380d0be1a7ad7552dc00c63/effdet/loss.py#L69)

### Alternatives

Wait for them to reach into fvcore or PyTorch. Till then we keep duplicating these code for models.

### Additional context

Note, if we are re-using implementation from any repo. Please cite them on top of code.

**Comment by hassiahk** [comment]
Do we include `sample_weight` and `reduction` parameters as well for the user to choose? Currently `dice_loss` does not consider `sample_weight` and takes `reduction = sum` as shown below.
https://github.com/Quick-AI/quickvision/blob/c6bd28e8d77905c6e4f9538644df5d875af654d2/quickvision/losses/segmentation.py#L6 https://github.com/Quick-AI/quickvision/blob/c6bd28e8d77905c6e4f9538644df5d875af654d2/quickvision/losses/segmentation.py#L21

Below is a part of `huber_loss` from the implementation you mentioned which considers both `sample_weight` and `reduction`.
```
def huber_loss(input, target, delta: float = 1., weights: Optional[torch.Tensor] = None, size_average: bool = True):
    # rest of the code
    if weights is not None:
        loss *= weights
    return loss.mean() if size_average else loss.sum()
```

I will start working on this simultaneously, if it is up for grabs. :smile:

**Comment by oke-aditya** [comment]
Yes ! I was about to tell about these.
We include `weights` and `reduction`. 

Reduction can be `none`, `sum` or `mean`.

P.S. We also keep consistent with torchvision and if it gets a losses API we will support only other losses.
We won't duplicate torchvision losses unless necessary.

**Comment by hassiahk** [comment]
What should be the structure for these losses? I can think of these three ways:
- Each loss will have a separate file.
- Club losses together in a single file based on the downstream task.
- Do it like how we did it for models (one single folder for a downstream task and one file for each loss).

**Comment by oke-aditya** [comment]
@zhiqwang you are welcome to contribute 😄

**Comment by zhiqwang** [comment]
Hi @oke-aditya , I'm looking for anything I can do in this vigorous repo 🚀

**Comment by oke-aditya** [comment]
Join on Slack [here](https://join.slack.com/t/quickai/shared_invite/zt-iz7tqk3r-IQa4SoxJGIK5WS8VdZhzeQ)

All the development talks go here ! You can freely communicate your ideas, RFCs and thoughts.

**Comment by oke-aditya** [comment]
Each loss in separate file is better This keeps abstraction minimal.

If We could create a folder called `nn` but that would interfere with `torch` which we can avoid.

Also, should we implement losses as Classes / functions ?

If we implement as `classes` we should inherit from `nn.Module`.
Otherwise we can continue to use them as functions.
This is slight confusion. I see no advantage of one over the other.
PyTorch creates them as classes and gives a functional API too.

Eg.
```
loss_fn = nn.CrossEntropyLoss()
loss_v = F.CrossEntropy(inp, op)
```
I think we should follow that and provide a functional API too (if possible)

Losses tied to models can be implemented as mixture of both. Since their use will be only with the model.
Preferabbly like `detr_loss`.


I propose to keep API Simple.
```
==>losses
====>functional
=======>loss_dice.py
====>__init__.py
====> loss_dice.py
====> loss_detr.py
```

Thoughts @zhiqwang  @hassiahk @ramaneswaran  ?

**Comment by hassiahk** [comment]
I agree with implementing losses as both `classes` and `functions` since it would be easier for people coming from `torch` to use them. We can just do this:
```
def loss_function(input, target):
    return loss

class LossFunction(nn.Module):
    def forward(self, input, target):
        return loss_function(input, target)
```

I did not get this part. How are we avoiding by naming the folder as `nn`?
> If We could create a folder called `nn` but that would interfere with `torch` which we can avoid.

**Comment by oke-aditya** [comment]
If we call our folder as `nn` people will have to do `from quickvision.nn import loss_fn`
While `nn` is something torch uses whose code is bonded to C++ API. So this doesn't serve this purpose.
Safer side we can keep losses.

**Comment by hassiahk** [comment]
Based on whatever we discussed above, I will start working on this and let's see how it goes.

**Comment by oke-aditya** [comment]
@zhiqwang @hassiahk I think we can start with this for release 2.0, Let;s propsose a template API for this.

**Comment by hassiahk** [comment]
@oke-aditya, are we implementing all the losses mentioned in the initial comment?

**Comment by oke-aditya** [comment]
Yes 😀
